sky/spot/core.py

```diff
@@ -0,0 +1,299 @@
+    """
+    entrypoint = task
+    dag_uuid = str(uuid.uuid4().hex[:4])
+
+    dag = dag_utils.convert_entrypoint_to_dag(entrypoint)
+    assert dag.is_chain(), ('Only single-task or chain DAG is '
+                            'allowed for spot_launch.', dag)
+
+    dag_utils.maybe_infer_and_fill_dag_and_task_names(dag)
+
+    task_names = set()
+    for task_ in dag.tasks:
+        if task_.name in task_names:
+            raise ValueError(
+                f'Task name {task_.name!r} is duplicated in the DAG. Either '
+                'change task names to be unique, or specify the DAG name only '
+                'and comment out the task names (so that they will be auto-'
+                'generated) .')
+        task_names.add(task_.name)
+
+    dag_utils.fill_default_spot_config_in_dag_for_spot_launch(dag)
+
+    for task_ in dag.tasks:
+        controller_utils.maybe_translate_local_file_mounts_and_sync_up(
+            task_, path='spot')
+
+    with tempfile.NamedTemporaryFile(prefix=f'spot-dag-{dag.name}-',
+                                     mode='w') as f:
+        dag_utils.dump_chain_dag_to_yaml(dag, f.name)
+        controller_name = spot_utils.SPOT_CONTROLLER_NAME
+        prefix = constants.SPOT_TASK_YAML_PREFIX
+        remote_user_yaml_path = f'{prefix}/{dag.name}-{dag_uuid}.yaml'
+        remote_user_config_path = f'{prefix}/{dag.name}-{dag_uuid}.config_yaml'
+        controller_resources = (controller_utils.get_controller_resources(
+            controller_type='spot',
+            controller_resources_config=constants.CONTROLLER_RESOURCES))
+
+        vars_to_fill = {
+            'remote_user_yaml_path': remote_user_yaml_path,
+            'user_yaml_path': f.name,
+            'spot_controller': controller_name,
+            # Note: actual spot cluster name will be <task.name>-<spot job ID>
+            'dag_name': dag.name,
+            'retry_until_up': retry_until_up,
+            'remote_user_config_path': remote_user_config_path,
+            'sky_python_cmd': skylet_constants.SKY_PYTHON_CMD,
+            'modified_catalogs':
+                service_catalog_common.get_modified_catalog_file_mounts(),
+            **controller_utils.shared_controller_vars_to_fill(
+                'spot',
+                remote_user_config_path=remote_user_config_path,
+            ),
+        }
+
+        yaml_path = os.path.join(constants.SPOT_CONTROLLER_YAML_PREFIX,
+                                 f'{name}-{dag_uuid}.yaml')
+        common_utils.fill_template(constants.SPOT_CONTROLLER_TEMPLATE,
+                                   vars_to_fill,
+                                   output_path=yaml_path)
+        controller_task = task_lib.Task.from_yaml(yaml_path)
+        assert len(controller_task.resources) == 1, controller_task
+        # Backward compatibility: if the user changed the
+        # spot-controller.yaml.j2 to customize the controller resources,
+        # we should use it.
```

Comment:
code gardening: when is this behaviour got changed? Could we remove it in this release?

Reply:
Good catch! This comment should be staled, since we have the feature to customize the controller resources in `~/.sky/confg.yaml` already. Removing now.